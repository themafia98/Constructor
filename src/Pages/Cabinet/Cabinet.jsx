import React,{Fragment} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Redirect} from 'react-router-dom';
import {CSSTransition} from 'react-transition-group';
/* ------- Event stream ------- */
import Events from 'events';
/* ------- HOC for het firebase controll object ------- */
import withFirebase from '../../components/withFirebase';
/* ------- Moddlewares ------- */
import {middlewareLogOutUser} from '../../redux/middleware/loadUserMiddleware';
import middlewareDelete from '../../redux/middleware/middlewareDelete';
/* ------- Child components ------- */
import Loader from '../../components/loading/Loader';
import Header from '../../components/header/Header';
import Modal from '../../components/modalWindow/ModalWindow';
import ProjectsSection from '../../components/ProjectsSection/ProjectsSection';
import './Cabinet.scss';

class Cabinet extends React.PureComponent {

	static propTypes = {
		firebase: PropTypes.object.isRequired, /** @firebase class for use firebase functions */
		active: PropTypes.bool, /** @active - status firebase auth */
		dispatch: PropTypes.func.isRequired, /** @dispatch - redux */
		history: PropTypes.object.isRequired, /** @Router HTML5 history */
		location: PropTypes.object.isRequired, /** @Router */
		match: PropTypes.object.isRequired, /** @Router */
		idUser: PropTypes.string, /** @Session user id from redux */
		projects: PropTypes.arrayOf(PropTypes.object).isRequired /** @Projects array with user projects from redux */
	}

	cabinetStream = new Events();

	state = {
		modalActive: false,
		workMode: 'default',
	}

	logOut = () => {
		/** destroy session */
		this.props.dispatch(middlewareLogOutUser(this.props.idUser));
	}

	deletItem = event => {
		/** destroy project */
		this.props.dispatch(middlewareDelete({...event, uid: this.props.firebase.getCurrentUser().uid}));
	};

		changeWorkMode = event => {
		/** workMode changes */
		this.setState ({
		...this.state,
		modalActive: event.active,
		workMode: event.action
		});
		}


	render(){

		if (this.props.active){
		return (
		<Fragment>
				<Header cabinetStream = {this.cabinetStream} 
					title = {this.props.config.title} 
					idUser = {this.props.idUser} 
				/>
				<div className = 'Cabinet'>
				<CSSTransition
					in={this.state.modalActive}
					timeout={300}
					classNames="modalAnimation"
					unmountOnExit
					appear
				>
				<Modal
					cabinetStream = {this.cabinetStream}
					workMode = {this.state.workMode} 
				/>
				</CSSTransition>
				<ProjectsSection cabinetStream = {this.cabinetStream} />
				</div>
			</Fragment>
		)
		} else if (!this.props.firebase.getCurrentUser()) return <Redirect to = { '/'} />
		else return <Loader path = '/img/loading.gif' type = 'Cabinet' />
	}


	componentDidMount = () => {
		this.cabinetStream.on('EventDeleteItem', this.deletItem);
		this.cabinetStream.on('EventChangeWorkMode', this.changeWorkMode);
	};

	componentWillUnmount = () => {
		this.cabinetStream.off('EventDeleteItem', this.deletItem);
		this.cabinetStream.off('EventChangeWorkMode', this.changeWorkMode);
	}
}

	const mapStateToProps = (state) => {
	return {
		idUser: state.cabinet.idUser,
		projects: [...state.cabinet.projects],
		active: state.cabinet.active
	}
	};

	export default connect(mapStateToProps)(withFirebase(Cabinet));
