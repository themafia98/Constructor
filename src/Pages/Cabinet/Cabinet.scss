.Cabinet {
    position: absolute;
    width: 100%;
    height: 100%;
    overflow: auto;

    div.cabinet_action_loader{
        position: absolute;
        width: 50px;
        height: 50px;
        top: 0;
        left: 0;
    }
}

.modalAnimation-enter {
    opacity: 0;
}
.modalAnimation-enter.modalAnimation-enter-active{
    opacity: 1;
    transition: opacity .5s;
}