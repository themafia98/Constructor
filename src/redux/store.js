import {createStore} from 'redux';
import combineReducers from './reducers';

let broswerPluginDev = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__();
const store = createStore(combineReducers,broswerPluginDev);
export default store;