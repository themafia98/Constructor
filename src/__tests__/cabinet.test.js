import React from 'react';
import firebase from 'firebase';
import {Provider} from 'react-redux';
import {MemoryRouter, Route} from 'react-router-dom';
import renderer from 'react-test-renderer';
import Cabinet from '../Pages/Cabinet/Cabinet';
import store from '../redux/testReduxStore';

const config = require('../config.json');

const props = {
    history: {},
    match: {},
    location: {}
}

const CabinetComponent = renderer.create(
    <MemoryRouter initialEntries = {['/Cabinet']}>
        <Route path = "/Cabinet" exact >
            <Provider store = {store}>
                <Cabinet   {...props} config = {config} firebase = {firebase} />
            </Provider>
        </Route>
    </MemoryRouter>
);

const act = renderer.act;
test("Cabinet test", function(){

    let snapshotCabient = CabinetComponent.toJSON();
    expect(snapshotCabient).toMatchSnapshot();
    act(() => {
    const controllerEnter = CabinetComponent.root.find(item => item.props.className === 'projectControllerButton_enter');

    controllerEnter.props.onClick({stopPropagation: () => null});
    });
    snapshotCabient = CabinetComponent.toJSON();
    expect(snapshotCabient).toMatchSnapshot();

    act(() => {
    const logoutButton = CabinetComponent.root.find(item => item.props.className === 'logoutButton');

    logoutButton.props.onClick({stopPropagation: () => null});
    });
    snapshotCabient = CabinetComponent.toJSON();
    expect(snapshotCabient).toMatchSnapshot();
    act(() => {
    const about = CabinetComponent.root.find(item => item.props.className === 'about');

    about.props.onClick({stopPropagation: () => null});
    });
    snapshotCabient = CabinetComponent.toJSON();
    expect(snapshotCabient).toMatchSnapshot();



});