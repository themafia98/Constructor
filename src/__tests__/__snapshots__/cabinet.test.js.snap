// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cabinet test 1`] = `
Array [
  <header>
    <div
      className="container"
    >
      <div
        className="flex-row"
      >
        <div
          className="header__CabinetInfo"
          onClick={[Function]}
        >
          <img
            alt="icon"
            className="mainIcon"
            draggable={false}
            onClick={null}
            src="/img/logo.svg"
            title={null}
          />
          <h3
            className="tilteApp"
          >
            Constructor
          </h3>
        </div>
        <div
          className="header__newProject__AddButton"
          onClick={[Function]}
        >
          <img
            alt="icon"
            className="Cabinet"
            draggable={false}
            onClick={null}
            src="/img/plus.png"
            title={null}
          />
        </div>
        <div
          className="Navigator"
        >
          <img
            alt="icon"
            className="about"
            draggable={false}
            onClick={[Function]}
            src="/img/about_logo.svg"
            title="about"
          />
          <img
            alt="icon"
            className="logoutButton"
            draggable={false}
            onClick={[Function]}
            src="/img/logOut.svg"
            title="logOut"
          />
        </div>
      </div>
    </div>
  </header>,
  <section
    className="projectsList"
  >
    <div
      className="container"
    >
      <div
        className="flex-column"
      >
        <div
          className="projectsList__about"
        >
          <h3>
            Your Projects
          </h3>
        </div>
        <div
          className="projectsList__list"
        >
          <div
            className="item"
            data-id={0}
          >
            <p
              className="ProjectName"
            >
              Test1
            </p>
            <p
              className="ProjectType"
            >
              landing
            </p>
            <div
              className="projectController"
            >
              <input
                className="projectControllerButton_enter"
                onClick={[Function]}
                type="button"
                value="enter build"
              />
              <input
                className="projectControllerButton_delete"
                onClick={[Function]}
                type="button"
                value="delete"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>,
]
`;
